lib/ansible/module_utils/common/validation.py

```diff
@@ -0,0 +1,242 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018 Ansible Project
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+from ansible.module_utils.common.collections import is_iterable
+from ansible.module_utils.six import string_types
+
+
+def count_terms(terms, module_parameters):
+    """Count the number of occurrences of a key in a given dictionary
+
+    :arg terms: String or iterable of values to check
+    :arg module_parameters: Dictionary of module parameters
+
+    :returns: An integer that is the number of occurrences of the terms values
+        in the provided dictionary.
+    """
+
+    if not is_iterable(terms):
+        terms = [terms]
+
+    return len(set(terms).intersection(module_parameters))
+
+
+def check_mutually_exclusive(terms, module_parameters):
+    """Check mutually exclusive terms against argument parameters. Accepts
+    a single list or list of lists that are groups of terms that should be
+    mutually exclusive with one another
+
+    :arg terms: List of mutually exclusive module parameters
+    :arg module_parameters: Dictionary of module parameters
+
+    :returns: A list of terms that are mutually exclusive.
+    """
+
+    result = []
+    if terms is None:
+        return result
+
+    for check in terms:
+        count = count_terms(check, module_parameters)
+        if count > 1:
+            result.append(check)
+
+    return result
+
+
+def check_required_one_of(terms, module_parameters):
+    """Check each list of terms to ensure at least one exists in the given module
```

Comment:
The same comment applies here.  I question the name vs what this is doing.  I feel like these `check_*` methods should be raising exceptions that are then caught in their `AnsibleModule` counterparts.  As it is, these are acting as "listing" functions, that then require the caller to know what a failure is based on what is returned.

As I am typing this I am still somewhat struggling, as it may be useful for the caller to also have the list.

I think in the end what I would like to see is an exception to indicate the failure, but with access in the exception to the list.

Reply:
I changed all these functions to raise an exception and pass the results as a parameter (along with the original parameters as well, though we probably don't need those) to the exception. That way it can be used by the caller catching the exception.